File: android/src/test/java/app/kaeru/domain/discover/SeasonTest.kt
package app.kaeru.domain.discover

import org.junit.Assert.assertEquals
import org.junit.Test
import java.time.Instant
import java.time.ZoneId
import java.time.ZoneOffset

class SeasonTest {
    private val utc = ZoneOffset.UTC

    private fun at(iso: String, zone: ZoneId = utc) = Season.current(Instant.parse(iso), zone)

    // --- current -------------------------------------------------------------------------------

    @Test
    fun `january through march is winter`() {
        assertEquals(Season(SeasonKind.WINTER, 2026), at("2026-01-01T00:00:00Z"))
        assertEquals(Season(SeasonKind.WINTER, 2026), at("2026-03-31T23:59:59Z"))
    }

    @Test
    fun `april through june is spring`() {
        assertEquals(Season(SeasonKind.SPRING, 2026), at("2026-04-01T00:00:00Z"))
        assertEquals(Season(SeasonKind.SPRING, 2026), at("2026-06-30T12:00:00Z"))
    }

    @Test
    fun `july through september is summer`() {
        assertEquals(Season(SeasonKind.SUMMER, 2026), at("2026-07-01T00:00:00Z"))
        assertEquals(Season(SeasonKind.SUMMER, 2026), at("2026-09-13T20:00:00Z"))
    }

    @Test
    fun `october through december is fall`() {
        assertEquals(Season(SeasonKind.FALL, 2026), at("2026-10-01T00:00:00Z"))
        assertEquals(Season(SeasonKind.FALL, 2026), at("2026-12-31T23:00:00Z"))
    }

    @Test
    fun `the season is the one the viewer is living in, not the one in UTC`() {
        // 22:00 on 31 December in London is already 11:00 on 1 January in Auckland.
        val newYear = Instant.parse("2026-12-31T22:00:00Z")
        assertEquals(Season(SeasonKind.FALL, 2026), Season.current(newYear, utc))
        assertEquals(Season(SeasonKind.WINTER, 2027), Season.current(newYear, ZoneId.of("Pacific/Auckland")))
    }

    // --- previous and next ---------------------------------------------------------------------

    @Test
    fun `previous walks back through the year`() {
        assertEquals(Season(SeasonKind.SPRING, 2026), Season(SeasonKind.SUMMER, 2026).previous())
        assertEquals(Season(SeasonKind.WINTER, 2026), Season(SeasonKind.SPRING, 2026).previous())
    }

    @Test
    fun `previous of winter is the autumn of the year before`() {
        assertEquals(Season(SeasonKind.FALL, 2025), Season(SeasonKind.WINTER, 2026).previous())
    }

    @Test
    fun `next walks forward through the year`() {
        assertEquals(Season(SeasonKind.SUMMER, 2026), Season(SeasonKind.SPRING, 2026).next())
        assertEquals(Season(SeasonKind.FALL, 2026), Season(SeasonKind.SUMMER, 2026).next())
    }

    @Test
    fun `next of autumn is the winter of the year after`() {
        assertEquals(Season(SeasonKind.WINTER, 2027), Season(SeasonKind.FALL, 2026).next())
    }

    @Test
    fun `stepping one way and back returns where it started`() {
        val summer = Season(SeasonKind.SUMMER, 2026)
        assertEquals(summer, summer.next().previous())
        assertEquals(summer, summer.previous().next())
    }

    // --- the wire ------------------------------------------------------------------------------

    @Test
    fun `apiValue is what Shikimori calls the season`() {
        assertEquals("winter_2026", Season(SeasonKind.WINTER, 2026).apiValue)
        assertEquals("spring_2026", Season(SeasonKind.SPRING, 2026).apiValue)
        assertEquals("summer_2026", Season(SeasonKind.SUMMER, 2026).apiValue)
        assertEquals("fall_2026", Season(SeasonKind.FALL, 2026).apiValue)
    }

    // --- the three choices ---------------------------------------------------------------------

    @Test
    fun `the choices are the season before, the one now, and the one coming`() {
        val summer = Season(SeasonKind.SUMMER, 2026)
        assertEquals(
            listOf(Season(SeasonKind.SPRING, 2026), summer, Season(SeasonKind.FALL, 2026)),
            seasonChoices(summer),
        )
    }

    @Test
    fun `the choices cross the new year without losing a season`() {
        assertEquals(
            listOf(Season(SeasonKind.FALL, 2025), Season(SeasonKind.WINTER, 2026), Season(SeasonKind.SPRING, 2026)),
            seasonChoices(Season(SeasonKind.WINTER, 2026)),
        )
    }
}

File: web/src/domain/season.test.ts
// Vectors from android/src/test/java/app/kaeru/domain/discover/SeasonTest.kt and the season part
// of android/src/test/java/app/kaeru/ui/common/design/FormatTest.kt.
import { describe, expect, it } from "vitest";
import type { Season } from "./season";
import { currentSeason, sameSeason, seasonApiValue, seasonChips, seasonTitle } from "./season";

// A local wall-clock instant, so the season is the viewer's whatever zone runs the tests.
function local(year: number, month: number, day: number, hour = 0, minute = 0, second = 0): number {
  return new Date(year, month - 1, day, hour, minute, second).getTime();
}

const season = (kind: Season["kind"], year: number): Season => ({ kind, year });

describe("currentSeason", () => {
  it("January through March is winter", () => {
    expect(currentSeason(local(2026, 1, 1))).toEqual(season("winter", 2026));
    expect(currentSeason(local(2026, 3, 31, 23, 59, 59))).toEqual(season("winter", 2026));
  });

  it("April through June is spring", () => {
    expect(currentSeason(local(2026, 4, 1))).toEqual(season("spring", 2026));
    expect(currentSeason(local(2026, 6, 30, 12))).toEqual(season("spring", 2026));
  });

  it("July through September is summer", () => {
    expect(currentSeason(local(2026, 7, 1))).toEqual(season("summer", 2026));
    expect(currentSeason(local(2026, 9, 13, 20))).toEqual(season("summer", 2026));
  });

  it("October through December is fall", () => {
    expect(currentSeason(local(2026, 10, 1))).toEqual(season("fall", 2026));
    expect(currentSeason(local(2026, 12, 31, 23))).toEqual(season("fall", 2026));
  });

  it("is the season of the local calendar across the new year", () => {
    expect(currentSeason(local(2026, 12, 31, 23, 59, 59))).toEqual(season("fall", 2026));
    expect(currentSeason(local(2027, 1, 1, 0, 0, 1))).toEqual(season("winter", 2027));
  });
});

describe("seasonChips", () => {
  it("offers the season before, the one now and the one coming", () => {
    expect(seasonChips(local(2026, 8, 1))).toEqual([season("spring", 2026), season("summer", 2026), season("fall", 2026)]);
    expect(seasonChips(local(2026, 5, 1))).toEqual([season("winter", 2026), season("spring", 2026), season("summer", 2026)]);
  });

  it("crosses the new year without losing a season", () => {
    expect(seasonChips(local(2026, 2, 1))).toEqual([season("fall", 2025), season("winter", 2026), season("spring", 2026)]);
    expect(seasonChips(local(2026, 11, 1))).toEqual([season("summer", 2026), season("fall", 2026), season("winter", 2027)]);
  });
});

describe("seasonApiValue", () => {
  it("is what Shikimori calls the season, with fall rather than autumn", () => {
    expect(seasonApiValue(season("winter", 2026))).toBe("winter_2026");
    expect(seasonApiValue(season("spring", 2026))).toBe("spring_2026");
    expect(seasonApiValue(season("summer", 2026))).toBe("summer_2026");
    expect(seasonApiValue(season("fall", 2026))).toBe("fall_2026");
  });
});

describe("seasonTitle", () => {
  it("names the season and the year", () => {
    expect(seasonTitle(season("winter", 2026))).toBe("Зима 2026");
    expect(seasonTitle(season("spring", 2026))).toBe("Весна 2026");
    expect(seasonTitle(season("summer", 2026))).toBe("Лето 2026");
    expect(seasonTitle(season("fall", 2026))).toBe("Осень 2026");
  });

  it("labels the three chips in sentence case with no separator", () => {
    const titles = seasonChips(local(2026, 10, 15)).map(seasonTitle);
    expect(titles).toEqual(["Лето 2026", "Осень 2026", "Зима 2027"]);
    for (const title of titles) {
      expect(title).not.toContain("·");
      expect(title).not.toBe(title.toUpperCase());
    }
  });
});

describe("sameSeason", () => {
  it("compares kind and year", () => {
    expect(sameSeason(season("fall", 2026), season("fall", 2026))).toBe(true);
    expect(sameSeason(season("fall", 2026), season("fall", 2025))).toBe(false);
    expect(sameSeason(season("fall", 2026), season("summer", 2026))).toBe(false);
  });
});
